helix-core/src/main/java/org/apache/helix/controller/stages/MessageDispatchStage.java

```diff
@@ -78,7 +79,17 @@ protected void processEvent(ClusterEvent event, MessageOutput messageOutput) thr
         batchMessage(dataAccessor.keyBuilder(), messagesToSend, resourceMap, liveInstanceMap,
             manager.getProperties());
 
+    // An early check for expected leader session. If the sessions don't match, it means the
+    // controller's session changes, then messages should not be sent and pipeline should stop.
+    Optional<String> expectedSession = event.getAttribute(AttributeName.EVENT_SESSION.name());
+    if (!expectedSession.isPresent() || !expectedSession.get().equals(manager.getSessionId())) {
+      throw new StageException(String.format(
```

Comment:
Can we skip the dispatch but continue the pipeline for the other stages?
I'm not sure, this is a question.